sigstore/_verify/verifier.py

```diff
@@ -0,0 +1,271 @@
+from sigstore._verify.models import (
+    VerificationFailure,
+    VerificationMaterials,
+    VerificationResult,
+    VerificationSuccess,
+)
+from sigstore._verify.policy import VerificationPolicy
+
+logger = logging.getLogger(__name__)
+
+
+_DEFAULT_FULCIO_ROOT_CERT = resources.read_binary("sigstore._store", "fulcio.crt.pem")
+_DEFAULT_FULCIO_INTERMEDIATE_CERT = resources.read_binary(
+    "sigstore._store", "fulcio_intermediate.crt.pem"
+)
+
+_STAGING_FULCIO_ROOT_CERT = resources.read_binary(
+    "sigstore._store", "fulcio.crt.staging.pem"
+)
+_STAGING_FULCIO_INTERMEDIATE_CERT = resources.read_binary(
+    "sigstore._store", "fulcio_intermediate.crt.staging.pem"
+)
+
+
+class RekorEntryMissing(VerificationFailure):
+    """
+    A specialization of `VerificationFailure` for Rekor lookup failures,
+    with additional lookup context.
+    """
+
+    reason: str = "Rekor has no entry for the given verification materials"
+    signature: str
+    artifact_hash: str
+
+
+class CertificateVerificationFailure(VerificationFailure):
+    """
+    A specialization of `VerificationFailure` for certificate signature
+    verification failures, with additional exception context.
+    """
+
+    reason: str = "Failed to verify signing certificate"
+    exception: Exception
+
+    class Config:
+        # Needed for the `exception` field above, since exceptions are
+        # not trivially serializable.
+        arbitrary_types_allowed = True
+
+
+class Verifier:
+    def __init__(self, *, rekor: RekorClient, fulcio_certificate_chain: List[bytes]):
+        """
+        Create a new `Verifier`.
+
+        `rekor` is a `RekorClient` capable of connecting to a Rekor instance
+        containing logs for the file(s) being verified.
+
+        `fulcio_certificate_chain` is a list of PEM-encoded X.509 certificates,
+        establishing the trust chain for the signing certificate and signature.
+        """
+        self._rekor = rekor
+
+        self._fulcio_certificate_chain: List[X509] = []
+        for parent_cert_pem in fulcio_certificate_chain:
+            parent_cert = load_pem_x509_certificate(parent_cert_pem)
+            parent_cert_ossl = X509.from_cryptography(parent_cert)
+            self._fulcio_certificate_chain.append(parent_cert_ossl)
+
+    @classmethod
+    def production(cls) -> Verifier:
+        return cls(
+            rekor=RekorClient.production(),
+            fulcio_certificate_chain=[
+                _DEFAULT_FULCIO_ROOT_CERT,
+                _DEFAULT_FULCIO_INTERMEDIATE_CERT,
+            ],
+        )
+
+    @classmethod
+    def staging(cls) -> Verifier:
+        return cls(
+            rekor=RekorClient.staging(),
+            fulcio_certificate_chain=[
+                _STAGING_FULCIO_ROOT_CERT,
+                _STAGING_FULCIO_INTERMEDIATE_CERT,
+            ],
+        )
+
+    def verify(
+        self,
+        materials: VerificationMaterials,
+        policy: VerificationPolicy,
```

Comment:
I sort of think this should be `Optional[VerificationPolicy]`. I understand that we want to guide users into using more rigorous checks during verification, but I think the right spot for that is at the CLI level (which you're already doing) rather than making the API less flexible.

This is really just enforcing that *something* gets checked, but will happily accept something not very meaningful like the workflow trigger by itself.

Happy to be convinced otherwise though.

Reply:
Yeah, I see your point here -- IMO we *do* want to keep `VerificationPolicy` itself mandatory, but we should offer an `UnsafeNullPolicy` or similar that (effectively) performs a no-op.

My reasoning there is that it reduces the number of type-states (no optional to worry about) while producing the same result (API users who *really* want to skip policy verification can do so). Thoughts?

Reply:
Yep, sounds good. I like that better because even the name is warning the user haha.

Reply:
f5c097b